### PR TITLE
Spell check

### PR DESCRIPTION
**Pull request by windymilla**
1. If user has selected text, spell check will just check the selected words (or part words if word is not wholly within selection). Note that GG1 Spell **Query** does not have this feature, but Spell **Check** does.

2. If word appears more times in the file than the threshold value, it is assumed to be a good spelling and so is not reported - this is the same as GG1's Spell **Query** but Spell **Check** does not have this.

.

### CODE REVIEW

**Comment by windymilla** [comment]
Question for @srjfoo:

If you run the spell checker in GG2, it lists the spelling errors, and chooses the first error (or the user can choose any error) which then selects that word in the main text window. If you decide you want to adjust the threshold and re-run, then now it only rechecks that one word (because it is selected and we have just made spell check only check the selection if there is one. 
That seems a bit annoying (at least it was to me) because it's almost certainly not what you intended. If you had previously selected a region to spell check, then you probably want to re-check that region. If you previously spell checked the whole file, you probably want to re-check the whole file.

Any good ideas of how to make it more helpful?

My best idea so far (which would need to wait until #351 is merged, with its "restore selection" feature) - probably more accurate to describe the comments below as a brain dump / random thoughts:
Save the selection (if any) immediately before doing the spell check. When using "re-run" do a "restore selection" immediately before re-running. The re-run would then work on the same section of text as the initial run. 
Disadvantage: Consider this case - user does spell check on whole file, then realizes they meant to only spell check a selection, so they make a selection and re-run. The reverse case (accidentally only spellchecking a selection, then attempting to re-run with no selection to spell check everything) would also go wrong. With the above idea, re-running would override their selection/de-selection.
Fix? If user makes a selection (including selecting nothing), then clear the #351 saved selection

**Comment by windymilla** [comment]
@srjfoo - I've pushed another commit that uses the "alternative selection"/"highlighting" idea instead of the main selection mechanism when highlighting a word due to a click in a checker tool (also in WF). The color (orange) is the same as GG1.
This fixes the "spell check a selection, then re-run" problem.

Question: Would it be better if the selected words in the dialog were also in orange? I tried it, and having lots of orange was a bit overwhelming, so I've currently left them in the default selection color. Here's a screen shot of all orange compared with leaving it blue.

image

image

**Comment by srjfoo** [comment]
For what it's worth, this is what I see, (I assume) because my selection is still active. So, somewhat different from what you see. However, I agree with you. I'd rather the highlighting in the spellcheck results stay the user's system default. (believe it or not, I'm not really all that fond of pink 😁)

image

---

I don't know if you're seeing this on Windows, but I am on macOS: When re-running spell check (whether for a changed threshold or a different selection), the count in the upper-left-hand corner is set to "0 Entries" between the old results display and the new. I doubt that anyone (after the first time) will misinterpret it, because the spell check does finish and you have a final count.

My first thought was to leave that area blank while the script is working, but it probably makes more sense to either display something like a `...` or the `working` message that's displayed in the main window. My thinking is that the PPer is probably going to be focused on the spell check widget, not on the bottom right-hand corner of the main text window. (You are, of course, welcome to ignore both suggestions -- or punt on the issue. 😁)

**Review by srjfoo — APPROVED** [review]

File: src/guiguts/highlight.py
Comment (srjfoo) [inline src/guiguts/highlight.py:36]: Aaah, this is why I see black text on orange in the main text window. :D 

Black looks fine on orange, which is a good reason to choose a color that black and white are equally readable on, because otherwise, fancy footwork would be needed with the default theme.

Comment (windymilla) [reply src/guiguts/highlight.py:36]: See additional color notes in #353 relating to "fancy footwork" for default theme.

**Comment by windymilla** [comment]
> the count in the upper-left-hand corner is set to "0 Entries" between the old results display and the new

I've added #364 to tackle this

**Pull request merged**
